GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs

```diff
@@ -0,0 +1,176 @@
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using UnityEngine;
+
+[assembly: InternalsVisibleTo("GooglePlayInstant.Tests.Editor.QuickDeploy")]
+
+namespace GooglePlayInstant.Editor.QuickDeploy
+{
+    /// <summary>
+    /// A class representing a server that provides an endpoint to use for getting authorization code from Google's
+    /// OAuth2 API's authorization page. On that page, the user grants the application to access their data on Google
+    /// Cloud platform. The server attempt to open and start listening at a fixed port. The server announces
+    /// its chosen endpoint with the CallbackEndpoint property. This server will run until it receives the first
+    /// request, which it will process to retrieve the authorization response and handle the by invoking on the response
+    /// the handler passed to the server during instatiation. The server will then stop listening for further requests.
+    ///
+    /// <see cref="https://developers.google.com/identity/protocols/OAuth2#installed"/> for an overview of OAuth2
+    /// protocol for installed applications that interact with Google APIs.
+    /// </summary>
+    public class OAuth2Server
+    {
+        private const string CloseTabText = "You may close this tab.";
+        private const int ServerPort = 2806;
+        private HttpListener _httpListener;
+        private string _callbackEndpoint;
+
+        private readonly Action<KeyValuePair<string, string>> _onResponseAction;
+
+        /// <summary>
+        /// An endpoint that on which the server is listening.
+        /// </summary>
+        public string CallbackEndpoint
+        {
+            get { return _callbackEndpoint; }
+        }
+
+        /// <summary>
+        /// An instance of a server that will run locally to retrieve authorization code. The server will stop running
+        /// once the first response gets received.
+        /// </summary>
+        /// <param name="onResponseAction">An action to be invoked on the key-value pair representing the first
+        /// response that will be caught by the server. Note that that the invocation of this handler will not be done
+        /// on the main thread, therefore the handler should only run operations that can be run on the main thread.
+        /// </param>
+        public OAuth2Server(Action<KeyValuePair<string, string>> onResponseAction)
+        {
+            _onResponseAction = onResponseAction;
+        }
+
+        /// <summary>
+        /// Starts this server to make it listen for requests containing authorization code or error data that are
+        /// forwarded from google's OAuth2 authorization url.
+        ///
+        /// After a call to this method, the CallBackEndpoint property of this instance will provide the endpoint
+        /// at which this server is listening.
+        /// </summary>
+        public void Start()
+        {
```

Comment:
Is it OK for Start() to be called more than once? If not, check _callbackEndpoint for not null and throw an exception.

If it is OK, check for _httpListener being null (indicating never started or Stop() called) and throw an exception if necessary.